lib/std/mem.zig

```diff
@@ -876,6 +874,63 @@ pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, nee
     return null;
 }
 
+fn boyerMooreHorspoolPreprocess(pattern: []const u8, table: []usize) void {
+    for (table) |*c| {
+        c.* = pattern.len;
+    }
+
+    var i: usize = 0;
+    while (i < pattern.len - 1) : (i += 1) {
+        table[pattern[i]] = pattern.len - 1 - i;
+    }
+}
+/// Find the index in a slice of a sub-slice, searching from the end backwards.
+/// To start looking at a different index, slice the haystack first.
+// Reverse boyer-moore-horspool algorithm
+pub fn lastIndexOf(comptime T: type, haystack: []const T, needle: []const T) ?usize {
+    if (haystack.len < 32 or needle.len <= 2)
+        return lastIndexOfNaive(T, haystack, needle);
+
+    if (needle.len > haystack.len or needle.len == 0) return null;
```

Comment:
`needle.len==0` will never reach this point. Maybe the `needle.len > haystack.len` can be factored out between this and `lastIndexOfNaive` as this is the only public-facing fn.
The `needle.len==0` fast path applies to both the fns.

Reply:
fixed